File: java/StudentList.java
public class StudentList {
    protected Student head;

    StudentList() {
        this.head = null;
    }

    public boolean empty(){
        return head == null;
    }

    public boolean isNpmUnique(Student data) {
        Student current = head;
        while (current != null) {
            if (current.getNpm().equals(data.getNpm())) {
                return false;
            }
            current = current.next;
        }
        return true;
    }
    
    public int size(){
        int count = 0;
        Student current = head;
        while(current != null){
            count++;
            current = current.next;
        }
        return count;
    }

    public boolean insertFirst(Student node){
        if(!this.isNpmUnique(node)){
            System.out.println("NPM sudah ada");
            return false;
        }
        node.next = head;
        head = node;
        return true;
    }

    public boolean insertLast(Student node){
        if(!this.isNpmUnique(node)){
            System.out.println("NPM sudah ada");
            return false;
        }
        Student current = head;
        if(current == null){
            head = node;
            return true;
        }
        else{
            while(current.next != null){
                current = current.next;
            }
            current.next = node;
            return true;
        }
    }

    public boolean insertOn(Student node, int position){
        if(!this.isNpmUnique(node)){
            System.out.println("NPM sudah ada");
            return false;
        }
        if(position < 0 || position > this.size()){
            return false;
        }
        if(position == 0){
            insertFirst(node);
            return true;
        }
        if(position == this.size()-1){
            insertLast(node);
            return true;
        }
        Student current = head;
        for(int i = 0; i < position - 1; i++){
            current = current.next;
        }
        node.next = current.next;
        current.next = node;
        return true;
    }

    public Student deleteFirst(){
        if(empty()){
            return null;
        }
        Student temp = head;
        head = head.next;
        return temp;
    }

    public Student deleteLast(){
        if(empty()){
            return null;
        }
        Student current = head;
        Student previous = null;
        while(current.next != null){
            previous = current;
            current = current.next;
        }
        previous.next = null;
        return current;
    }

    public Student deleteOn(int position){
        if(position < 0 || position > this.size()){
            return null;
        }
        if(position == 0){
            return deleteFirst();
        }
        if(position == this.size()-1){
            return deleteLast();
        }
        Student current = head;
        for(int i = 0; i < position - 1; i++){
            current = current.next;
        }
        Student temp = current.next;
        current.next = current.next.next;
        return temp;
    }

    public Student deleteByNpm(String npm){
        Student current = head;
        Student previous = null;
        while(current != null){
            if(current.getNpm().equals(npm)){
                if(previous == null){
                    head = head.next;
                }
                else{
                    previous.next = current.next;
                }
                return current;
            }
            previous = current;
            current = current.next;
        }
        return null;
    }

    public void printTable(){
        System.out.printf("%-15s %-5s %-5s %-10s %-10s %-10s %-10s %-10s %-10s\n", 
        "Nama", "Umur", "NPM", "Penguji1", "Penguji2", "Pembimbing", "NilaiAkhir", "NilaiMutu", "Status");
        for(Student temp = head; temp != null; temp = temp.next){
            System.out.printf("%-15s %-5d %-5s %-10s %-10s %-10s %-10s %-10s %-10s\n", temp.getName(), temp.getAge(), temp.getNpm(), temp.getNilaiPenguji1(), temp.getNilaiPenguji2(), temp.getNilaiPembimbing(), temp.getNilaiAkhir(), temp.getHurufMutu(), temp.getStatus());
        }
    }

    public void printFilteredTable(int sel, float nilai){
        System.out.printf("%-15s %-5s %-5s %-10s %-10s %-10s %-10s %-10s %-10s\n", 
        "Nama", "Umur", "NPM", "Penguji1", "Penguji2", "Pembimbing", "NilaiAkhir", "NilaiMutu", "Status");
        if(sel == 1){
            for(Student temp = head; temp != null; temp = temp.next){
                if(temp.getNilaiAkhir() > nilai){
                    System.out.printf("%-15s %-5d %-5s %-10s %-10s %-10s %-10s %-10s %-10s\n", temp.getName(), temp.getAge(), temp.getNpm(), temp.getNilaiPenguji1(), temp.getNilaiPenguji2(), temp.getNilaiPembimbing(), temp.getNilaiAkhir(), temp.getHurufMutu(), temp.getStatus());
                }
            }
        }else{
            for(Student temp = head; temp != null; temp = temp.next){
                if(temp.getNilaiAkhir() < nilai){
                    System.out.printf("%-15s %-5d %-5s %-10s %-10s %-10s %-10s %-10s %-10s\n", temp.getName(), temp.getAge(), temp.getNpm(), temp.getNilaiPenguji1(), temp.getNilaiPenguji2(), temp.getNilaiPembimbing(), temp.getNilaiAkhir(), temp.getHurufMutu(), temp.getStatus());
                }
            }
        }
    }
}

File: python/StudentList.py
# /*
#   Nama File : StudentList.py
#   Nama      : Andaru Danurdara Wibisana, Fauzan Azmi Dwicahyo, Irfan Kamal, Muhammad Attila An Naufal
#   NPM       : 140810200020, 140810200030, 140810200032, 140810200048
#   Kelas     : B
#   Tanggal   : 17 November 2021
#   Deskripsi : Implementasi kasus linked list nilai sidang menggunakan C++, class StudentList
# *********************************************/

from Student import Student
class StudentList(Student):
    def __init__(self):
        self.first = None

    def empty(self):
        if self.first is None:
            return True
        else: 
            return False
    
    def isNpmUnique(self, node):
        current = self.first
        while current is not None:
            if current.getNpm() == node.getNpm():
                return False
            current = current.next
        return True

    def size(self):
        count = 0
        current = self.first
        while current is not None:
            count += 1
            current = current.next
        return count

    def insertFirst(self, node):
        if self.isNpmUnique(node) == False:
            return False
        node.next = self.first
        self.first = node
        return True
    
    def insertLast(self, node):
        if self.isNpmUnique(node) == False:
            return False
        current = self.first
        if current is None:
            self.first = node
            return True
        else:
            while current.next is not None:
                current = current.next
            current.next = node
            return True

    def insertOn(self, node, position):
        if self.isNpmUnique(node) == False:
            return False
        if position < 0 or position > self.size():
            return False
        if position == 0:
            self.insertFirst(node)
            return True
        if position == self.size():
            self.insertLast(node)
            return True
        current = self.first
        for i in range (0, position-1):
            current = current.next
        node.next = current.next
        current.next = node
        return True
    
    def deleteFirst(self):
        if self.empty() :
            return None
        temp = self.first
        self.first = self.first.next
        return temp

    def deleteLast(self):
        if self.empty():
            return None
        current = self.first
        previous = None
        while current.next is not None:
            previous = current
            current = current.next
        previous.next = None
        return current
    
    def deleteOn(self, position):
        if position < 0 or position > self.size():
            return None
        if position == 0:
            return self.deleteFirst
        if position == self.size():
            return self.deleteLast
        current = self.first
        for i in range(0, position-1):
            current = current.next
        temp = current.next
        current.next = temp.next
        return temp

    def deleteByNpm(self, npm):
        current = self.first
        previous = None
        while current is not None:
            if current.getNpm() == npm:
                if previous is None :
                    self.first = self.first.next
                else:
                    previous.next = current.next
                return current
            previous = current
            current = current.next
        return None
    
    def printTable(self):
        print("Nama\tUmur\tNPM\tPenguji1\tPenguji2\tPembimbing\tNilaiAkhir\tNilaiMutu\tStatus")
        temp = self.first
        while temp is not None:
            print(temp.getName() + "\t" + str(temp.getAge()) + "\t" + str(temp.getNpm()) + "\t" + str(temp.getNilaiPenguji1()) + "\t\t" + str(temp.getNilaiPenguji2()) + "\t\t" + str(temp.getNilaiPembimbing()) + "\t\t" + str(temp.getNilaiAkhir()) + "\t\t" + temp.getHurufMutu() + "\t\t" + temp.getStatus())
            temp = temp.next
    
    def printFilteredTable(self, sel, nilai):
        print("Nama\tUmur\tNPM\tPenguji1\tPenguji2\tPembimbing\tNilaiAkhir\tNilaiMutu\tStatus")
        if sel == '1':
            temp = self.first
            while temp is not None:
                if temp.getNilaiAkhir() > nilai:
                    print(temp.getName() + "\t" + str(temp.getAge()) + "\t" + str(temp.getNpm()) + "\t" + str(temp.getNilaiPenguji1()) + "\t\t" + str(temp.getNilaiPenguji2()) + "\t\t" + str(temp.getNilaiPembimbing()) + "\t\t" + str(temp.getNilaiAkhir()) + "\t\t" + temp.getHurufMutu() + "\t\t" + temp.getStatus())
                temp = temp.next
        else:
            temp = self.first
            while temp is not None:
                if temp.getNilaiAkhir() < nilai:
                    print(temp.getName() + "\t" + str(temp.getAge()) + "\t" + str(temp.getNpm()) + "\t" + str(temp.getNilaiPenguji1()) + "\t\t" + str(temp.getNilaiPenguji2()) + "\t\t" + str(temp.getNilaiPembimbing()) + "\t\t" + str(temp.getNilaiAkhir()) + "\t\t" + temp.getHurufMutu() + "\t\t" + temp.getStatus())
                temp = temp.next